File: source/com/barrybecker4/ui1/gradient/RoundGradientPaint.java
package com.barrybecker4.ui1.gradient;

import java.awt.*;
import java.awt.geom.AffineTransform;
import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;
import java.awt.image.ColorModel;

/**
 * Source derived from Java 2D graphics book by J. Knudsen.
 */
public class RoundGradientPaint
        implements Paint {
    protected Point2D point;
    protected Point2D radius;
    protected Color pointColor, bgColor;

    public RoundGradientPaint( double x, double y, Color pointColor,
                               Point2D radius, Color backgroundColor ) {
        if ( radius.distance( 0, 0 ) <= 0 ) {
            throw new IllegalArgumentException( "Radius must be greater than 0." );
        }
        point = new Point2D.Double( x, y );
        this.pointColor = pointColor;
        this.radius = radius;
        bgColor = backgroundColor;
    }

    public PaintContext createContext(ColorModel cm,
                                      Rectangle deviceBounds, Rectangle2D userBounds,
                                      AffineTransform xform, RenderingHints hints ) {
        Point2D transformedPoint = xform.transform(point, null );
        Point2D transformedRadius = xform.deltaTransform(radius, null );
        return new RoundGradientContext( transformedPoint, pointColor,
                transformedRadius, bgColor);
    }

    public int getTransparency() {
        int a1 = pointColor.getAlpha();
        int a2 = bgColor.getAlpha();
        return (((a1 & a2) == 0xff) ? OPAQUE : TRANSLUCENT);
    }
}

File: scala-source/com/barrybecker4/ui/gradient/RoundGradientPaint.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.ui.gradient

import java.awt._
import java.awt.geom.AffineTransform
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D
import java.awt.image.ColorModel
import java.awt.Transparency._


/**
  * Source derived from Java 2D graphics book by J. Knudsen.
  */
class RoundGradientPaint(val x: Double, val y: Double, var pointColor: Color, var radius: Point2D, var bgColor: Color)
  extends Paint {

  if (radius.distance(0, 0) <= 0) throw new IllegalArgumentException("Radius must be greater than 0.")
  protected var point = new Point2D.Double(x, y)

  override def createContext(cm: ColorModel, deviceBounds: Rectangle, userBounds: Rectangle2D,
                             xform: AffineTransform, hints: RenderingHints): PaintContext = {
    val transformedPoint = xform.transform(point, null)
    val transformedRadius = xform.deltaTransform(radius, null)
    new RoundGradientContext(transformedPoint, pointColor, transformedRadius, bgColor)
  }

  override def getTransparency: Int = {
    val a1 = pointColor.getAlpha
    val a2 = bgColor.getAlpha
    if ((a1 & a2) == 0xff) OPAQUE else TRANSLUCENT
  }
}